docs/tutorials/streaming_analytics.md

```diff
@@ -0,0 +1,460 @@
+
+input
+    .keyBy(x -> x.key)
+    .window(TumblingEventTimeWindows.of(Time.minutes(1)))
+    .process(new MyWastefulMax());
+
+public static class MyWastefulMax extends ProcessWindowFunction<
+        SensorReading,                  // input type
+        Tuple3<String, Long, Integer>,  // output type
+        String,                         // key type
+        TimeWindow> {                   // window type
+    
+    @Override
+    public void process(
+            String key,
+            Context context, 
+            Iterable<SensorReading> events,
+            Collector<Tuple3<String, Long, Integer>> out) {
+
+        int max = 0;
+        for (SensorReading event : events) {
+            if (event.value > max) max = event.value;
+        }
+        out.collect(new Tuple3<>(key, context.window().getEnd(), max));
+    }
+}
+{% endhighlight %}
+
+A couple of things to note in this implementation:
+
+* All of the events assigned to the window have to be buffered in keyed Flink state until the window
+  is triggered. This is potentially quite expensive.
+* Our ProcessWindowFunction is being passed a Context object from which contains information about
+  the window. Its interface looks like this:
+
+{% highlight java %}
+public abstract class Context implements java.io.Serializable {
+    public abstract W window();
+    
+    public abstract long currentProcessingTime();
+    public abstract long currentWatermark();
+
+    public abstract KeyedStateStore windowState();
+    public abstract KeyedStateStore globalState();
+}
+{% endhighlight %}
+
+`windowState` and `globalState` are places where you can store per-key, per-window, or global
+per-key information. This might be useful, for example, if you want to record something about the
+current window and use that when processing a subsequent window.
+
+#### Incremental Aggregation Example
+
+{% highlight java %}
+DataStream<SensorReading> input = ...
+
+input
+    .keyBy(x -> x.key)
+    .window(TumblingEventTimeWindows.of(Time.minutes(1)))
+    .reduce(new MyReducingMax(), new MyWindowFunction());
+
+private static class MyReducingMax implements ReduceFunction<SensorReading> {
+    public SensorReading reduce(SensorReading r1, SensorReading r2) {
+        return r1.value() > r2.value() ? r1 : r2;
+    }
+}
+
+private static class MyWindowFunction extends ProcessWindowFunction<
+    SensorReading, Tuple3<String, Long, SensorReading>, String, TimeWindow> {
+
+    @Override
+    public void process(
+            String key,
+            Context context,
+            Iterable<SensorReading> maxReading,
+            Collector<Tuple3<String, Long, SensorReading>> out) {
+
+        SensorReading max = maxReading.iterator().next();
+        out.collect(new Tuple3<String, Long, SensorReading>(key, context.window().getEnd(), max));
+    }
+}
+{% endhighlight %}
+
+This implementation uses a more robust `KeySelector`. Notice also that the `Iterable<SensorReading>`
```

Comment:
The `KeySelector` is actually the same...

Reply:
Oops.